File: apps/mobile/test/analytics_test.dart
// ============================================================
// AGRITRADE STAGE 9: MOBILE FARM ANALYTICS & INSIGHT TESTS
// Unit and Widget verification for deterministic intelligence
// ============================================================

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:agro_product_app/features/analytics/application/analytics_calculator.dart';
import 'package:agro_product_app/features/analytics/application/farm_insight_engine.dart';
import 'package:agro_product_app/features/analytics/domain/analytics_metrics.dart';
import 'package:agro_product_app/features/analytics/domain/farm_insight.dart';
import 'package:agro_product_app/features/analytics/presentation/farm_insights_screen.dart';
import 'package:agro_product_app/features/analytics/presentation/providers/analytics_providers.dart';
import 'package:agro_product_app/features/analytics/presentation/widgets/insight_detail_sheet.dart';

void main() {
  group('Stage 9: AnalyticsCalculator Math Verification', () {
    test('calculates total savings correctly across discounts, delivery, and bulk', () {
      final total = AnalyticsCalculator.calculateTotalSavings(6840.0, 1200.0, 600.0);
      expect(total, 8640.0);
    });

    test('calculates traditional retail estimate with 16.5% offline markup', () {
      final retail = AnalyticsCalculator.calculateRetailComparison(46200.0, 16.5);
      expect(retail, 53823.0);
    });

    test('calculates average realized discount percentage', () {
      final discountPct = AnalyticsCalculator.calculateAverageDiscount(46200.0, 8640.0);
      expect(discountPct, 15.8);
    });

    test('calculates delivery savings vs rural freight standard', () {
      final deliverySavings = AnalyticsCalculator.calculateDeliverySavings(12, 600.0);
      // 12 * 150 = 1800 - 600 = 1200
      expect(deliverySavings, 1200.0);
    });

    test('computes spending change direction and delta percentage', () {
      final up = AnalyticsCalculator.calculateSpendingChange(46200.0, 39800.0);
      expect(up['trendDirection'], 'up');
      expect(up['percentageChange'], 16.1);

      final down = AnalyticsCalculator.calculateSpendingChange(12000.0, 15000.0);
      expect(down['trendDirection'], 'down');
      expect(down['percentageChange'], 20.0);

      final flat = AnalyticsCalculator.calculateSpendingChange(10000.0, 10020.0);
      expect(flat['trendDirection'], 'flat');
      expect(flat['percentageChange'], 0.0);
    });

    test('normalizes category allocation shares to 100%', () {
      const cats = [
        CategorySpend(
          categoryId: 'c1',
          categoryName: 'Fertilizers',
          amount: 5000.0,
          percentage: 0.0,
          orderCount: 2,
          topItem: 'NPK',
          growthRate: 25.0,
        ),
        CategorySpend(
          categoryId: 'c2',
          categoryName: 'Seeds',
          amount: 3000.0,
          percentage: 0.0,
          orderCount: 1,
          topItem: 'Soybean',
          growthRate: 0.0,
        ),
        CategorySpend(
          categoryId: 'c3',
          categoryName: 'Crop Protection',
          amount: 2000.0,
          percentage: 0.0,
          orderCount: 1,
          topItem: 'Coragen',
          growthRate: 33.0,
        ),
      ];

      final normalized = AnalyticsCalculator.normalizeCategoryPercentages(cats);
      expect(normalized[0].percentage, 50.0);
      expect(normalized[1].percentage, 30.0);
      expect(normalized[2].percentage, 20.0);
      final sum = normalized.fold<double>(0.0, (acc, c) => acc + c.percentage);
      expect(sum, 100.0);
    });
  });

  group('Stage 9: FarmInsightEngine Deterministic Evaluation', () {
    test('generates transparent, explainable decision insights from snapshot', () {
      final notifier = FarmAnalyticsNotifier();
      final snapshot = notifier.state;
      final insights = FarmInsightEngine.evaluate(snapshot);

      expect(insights.isNotEmpty, isTrue);

      // Verify spending anomaly rule
      final anomaly = insights.where((i) => i.type == InsightType.spendingAnomaly);
      expect(anomaly.isNotEmpty, isTrue);
      expect(anomaly.first.badgeLabel, 'Purchase Pattern');

      // Verify market opportunity rule
      final marketSignal = insights.where((i) => i.type == InsightType.priceOpportunity);
      expect(marketSignal.isNotEmpty, isTrue);
      expect(marketSignal.first.badgeLabel, 'Market Signal');

      // Verify delivery SLA rule
      final deliverySignal = insights.where((i) => i.type == InsightType.deliveryPerformance);
      expect(deliverySignal.isNotEmpty, isTrue);
      expect(deliverySignal.first.badgeLabel, 'Delivery Performance');

      // Verify transparency fields are populated on all insights
      for (final ins in insights) {
        expect(ins.whatHappened.length, greaterThan(15));
        expect(ins.whyDetected.length, greaterThan(15));
        expect(ins.supportingData.isNotEmpty, isTrue);
        expect(ins.recommendedConsideration.length, greaterThan(15));
      }
    });

    test('updates snapshot metrics dynamically when time range changes', () {
      final notifier = FarmAnalyticsNotifier();
      expect(notifier.state.timeRange, TimeRangeEnum.thirtyDays);

      notifier.setTimeRange(TimeRangeEnum.sevenDays);
      expect(notifier.state.timeRange, TimeRangeEnum.sevenDays);
      expect(notifier.state.spending.totalSpend, 14850.0);

      notifier.setTimeRange(TimeRangeEnum.threeMonths);
      expect(notifier.state.timeRange, TimeRangeEnum.threeMonths);
      expect(notifier.state.spending.totalSpend, 124300.0);

      notifier.setTimeRange(TimeRangeEnum.oneYear);
      expect(notifier.state.timeRange, TimeRangeEnum.oneYear);
      expect(notifier.state.spending.totalSpend, 342500.0);
    });
  });

  group('Stage 9: Mobile UI Widget Tests', () {
    testWidgets('renders FarmInsightsScreen with header, spend card, and categories', (tester) async {
      await tester.pumpWidget(
        const ProviderScope(
          child: MaterialApp(
            home: FarmInsightsScreen(),
          ),
        ),
      );
      await tester.pumpAndSettle();

      expect(find.text('Farm Insights'), findsOneWidget);
      expect(find.text('TOTAL AGRICULTURAL OUTLAY'), findsOneWidget);
      expect(find.text('INPUT CATEGORY ALLOCATION'), findsOneWidget);
      expect(find.text('FARMER SAVINGS INTELLIGENCE'), findsOneWidget);
      expect(find.text('LOGISTICS SLA & TURNAROUND'), findsOneWidget);
      expect(find.text('APMC COMMODITY BENCHMARKS'), findsOneWidget);
    });

    testWidgets('renders InsightDetailSheet with complete diagnostic transparency', (tester) async {
      const testInsight = FarmInsight(
        id: 'ins_test_1',
        type: InsightType.spendingAnomaly,
        severity: InsightSeverity.warning,
        badgeLabel: 'Purchase Pattern',
        title: 'Crop Protection spend is +43.8% above prior period',
        summary: 'Pesticide procurement increased significantly.',
        detectedAt: 'Deterministic Audit',
        supportingMetric: '+43.8% vs Baseline',
        whatHappened: 'Pesticide purchases increased following rain showers.',
        whyDetected: 'Deterministic threshold flagged 43.8% growth vs standard 15% limit.',
        supportingData: [
          InsightDataRow(metric: 'Current Spend', value: '₹12,800', benchmark: '₹8,900'),
          InsightDataRow(metric: 'Budget Share', value: '27.7%', benchmark: '30% max'),
        ],
        recommendedConsideration: 'Audit field application rates with agronomist before reordering.',
      );

      await tester.pumpWidget(
        const MaterialApp(
          home: Scaffold(
            body: InsightDetailSheet(insight: testInsight),
          ),
        ),
      );
      await tester.pumpAndSettle();

      expect(find.text('PURCHASE PATTERN'), findsOneWidget);
      expect(find.text('Crop Protection spend is +43.8% above prior period'), findsOneWidget);
      expect(find.text('WHAT HAPPENED'), findsOneWidget);
      expect(find.text('WHY DETECTED (DETERMINISTIC AUDIT)'), findsOneWidget);
      expect(find.text('SUPPORTING VERIFICATION DATA'), findsOneWidget);
      expect(find.text('RECOMMENDED ACTION / CONSIDERATION'), findsOneWidget);
      expect(find.text('Acknowledge Insight'), findsOneWidget);
    });
  });
}

File: apps/web/test/analytics.test.mjs
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

// Deterministic Calculation Logic matching features/analytics/application/aggregation-service.ts
function calculateTotalSavings(discountSavings, deliverySavings, bulkSavings = 0) {
  return Math.max(0, discountSavings + deliverySavings + bulkSavings);
}

function calculateRetailComparison(totalSpend, markupPercentage = 16.5) {
  if (totalSpend <= 0) return 0;
  return Math.round(totalSpend * (1 + markupPercentage / 100));
}

function calculateAverageDiscount(subtotal, discountAmount) {
  if (subtotal <= 0) return 0;
  return Math.min(100, Math.round((discountAmount / (subtotal + discountAmount)) * 1000) / 10);
}

function calculateDeliverySavings(deliveredOrderCount, actualDeliveryFeesPaid) {
  const standardRuralFreightBaseline = deliveredOrderCount * 150;
  return Math.max(0, standardRuralFreightBaseline - actualDeliveryFeesPaid);
}

function calculateSpendingChange(currentSpend, previousSpend) {
  if (previousSpend <= 0) {
    return {
      percentageChange: currentSpend > 0 ? 100 : 0,
      trendDirection: currentSpend > 0 ? 'up' : 'flat',
    };
  }
  const change = ((currentSpend - previousSpend) / previousSpend) * 100;
  const rounded = Math.round(change * 10) / 10;
  if (Math.abs(rounded) < 0.5) {
    return { percentageChange: 0, trendDirection: 'flat' };
  }
  return {
    percentageChange: Math.abs(rounded),
    trendDirection: rounded > 0 ? 'up' : 'down',
  };
}

function normalizeCategoryPercentages(categories) {
  const totalAmount = categories.reduce((sum, c) => sum + c.amount, 0);
  if (totalAmount <= 0) return categories;
  return categories.map((cat) => ({
    ...cat,
    percentage: Math.round((cat.amount / totalAmount) * 1000) / 10,
  }));
}

function getTimeRangeDays(range) {
  switch (range) {
    case '7d': return 7;
    case '30d': return 30;
    case '3m': return 90;
    case '6m': return 180;
    case '1y': return 365;
    default: return 30;
  }
}

// Deterministic Decision Insight Engine matching features/analytics/application/insight-engine.ts
class FarmInsightEngine {
  static evaluateSnapshot(snapshot) {
    const insights = [];

    // Rule 1: Spending Anomaly (>15% growth)
    const anomalousCategory = snapshot.categories.find(
      (cat) => cat.growthRate >= 15.0 && cat.amount > 3000
    );

    if (anomalousCategory) {
      insights.push({
        id: `ins_anomaly_${anomalousCategory.categoryId}_${snapshot.timeRange}`,
        type: 'spending_anomaly',
        severity: 'warning',
        category: 'spending',
        badgeLabel: 'Purchase Pattern',
        title: `${anomalousCategory.categoryName} spending is ${anomalousCategory.growthRate.toFixed(1)}% above previous baseline`,
        summary: `Capital allocation toward ${anomalousCategory.categoryName.toLowerCase()} reached ₹${anomalousCategory.amount.toLocaleString('en-IN')}, driven by recurring procurement of ${anomalousCategory.topItem}.`,
        detectedAt: 'Realtime Audit',
        supportingMetric: `+${anomalousCategory.growthRate.toFixed(1)}% vs Prior Period`,
        detail: {
          whatHappened: `Your spending on ${anomalousCategory.categoryName} increased from ₹${anomalousCategory.previousPeriodAmount.toLocaleString('en-IN')} to ₹${anomalousCategory.amount.toLocaleString('en-IN')}.`,
          whyDetected: `The AgriTrade deterministic threshold flagged an expenditure acceleration of ${anomalousCategory.growthRate.toFixed(1)}%, exceeding the normal operating limit of 15.0%.`,
          supportingData: [
            { metric: 'Current Period Spend', value: `₹${anomalousCategory.amount}`, benchmark: `₹${anomalousCategory.previousPeriodAmount}` },
            { metric: 'Budget Share', value: `${anomalousCategory.percentage}%`, benchmark: '30.0% ceiling' },
          ],
          recommendedConsideration: 'Audit field application rates with your regional agronomist.',
        },
      });
    }

    // Rule 2: Market Price Opportunity
    const discountedCommodity = snapshot.marketComparisons?.find(
      (c) => c.movementPercent <= -5.0 || (c.avgPrice - c.currentPrice) >= 80
    );

    if (discountedCommodity) {
      insights.push({
        id: `ins_mkt_${discountedCommodity.commodity.toLowerCase()}_${snapshot.timeRange}`,
        type: 'price_opportunity',
        severity: 'info',
        category: 'market',
        badgeLabel: 'Market Signal',
        title: `${discountedCommodity.commodity} trading below 30-day APMC average`,
        summary: `Spot rate at ${discountedCommodity.mandi} is currently ₹${discountedCommodity.currentPrice.toLocaleString('en-IN')}.`,
        detectedAt: 'Live APMC Feed',
        supportingMetric: `₹${discountedCommodity.currentPrice} ${discountedCommodity.unit}`,
        detail: {
          whatHappened: `${discountedCommodity.commodity} price has softened by ${Math.abs(discountedCommodity.movementPercent).toFixed(1)}% at ${discountedCommodity.mandi}.`,
          whyDetected: `Observed spot price is at ₹${discountedCommodity.currentPrice}, compared against an average of ₹${discountedCommodity.avgPrice}.`,
          supportingData: [
            { metric: 'Spot Modal Price', value: `₹${discountedCommodity.currentPrice}`, benchmark: `₹${discountedCommodity.avgPrice}` },
            { metric: '30-Day Range', value: `₹${discountedCommodity.minPrice} – ₹${discountedCommodity.maxPrice}` },
          ],
          recommendedConsideration: 'Evaluate holding stock in accredited warehouse storage until arrivals stabilize.',
        },
      });
    }

    // Rule 3: Delivery Performance Reliability
    const logistics = snapshot.deliveryPerformance;
    if (logistics && logistics.onTimeRate >= 96.0 && logistics.totalOrders > 0) {
      insights.push({
        id: `ins_logistics_${snapshot.timeRange}`,
        type: 'delivery_performance',
        severity: 'positive',
        category: 'logistics',
        badgeLabel: 'Delivery Performance',
        title: `${logistics.onTimeRate.toFixed(1)}% on-time fulfillment rate across ${logistics.totalOrders} shipments`,
        summary: `Fulfillment averaging ${logistics.averageDeliveryHours} hours with ${logistics.deliveryAttemptRate.toFixed(1)}% first-attempt success rate.`,
        detectedAt: 'Telematics Telemetry',
        supportingMetric: `${logistics.averageDeliveryHours}h Avg Turnaround`,
        detail: {
          whatHappened: `Across ${logistics.totalOrders} total orders, logistics partners sustained ${logistics.onTimeRate.toFixed(1)}% on-time delivery compliance.`,
          whyDetected: 'Stage 8 rural hub routing and local cargo EV telematics maintained transit timelines within promised delivery windows.',
          supportingData: [
            { metric: 'On-Time Fulfillment', value: `${logistics.onTimeRate.toFixed(1)}%`, benchmark: '95.0% Institutional SLA' },
            { metric: 'Average Transit Duration', value: `${logistics.averageDeliveryHours} Hours`, benchmark: '48h Rural Standard' },
          ],
          recommendedConsideration: 'Your delivery lane has verified rural access. Keep delivery pin synchronized.',
        },
      });
    }

    // Rule 4: Seasonal Planning Reminder
    const activePattern = snapshot.seasonalPatterns?.[0];
    if (activePattern) {
      insights.push({
        id: `ins_season_${activePattern.id}_${snapshot.timeRange}`,
        type: 'seasonal_reminder',
        severity: 'info',
        category: 'planning',
        badgeLabel: 'Farm Insight',
        title: `Seasonal input cycle: ${activePattern.season} window active`,
        summary: activePattern.observation,
        detectedAt: 'Agronomic Calendar',
        supportingMetric: activePattern.peakMonths,
        detail: {
          whatHappened: `Historical purchasing patterns show high concentration in ${activePattern.primaryCategory} during ${activePattern.peakMonths}.`,
          whyDetected: `Aggregated data indicates ${activePattern.season} seasonal operations generate repeatable input demand patterns with ${activePattern.confidenceScore}% historical recurrence.`,
          supportingData: [
            { metric: 'Primary Input Category', value: activePattern.primaryCategory },
            { metric: 'Historical Seasonal Outlay', value: `₹${activePattern.historicalSpend}` },
          ],
          recommendedConsideration: activePattern.recommendedAction,
        },
      });
    }

    return insights;
  }
}

// Deterministic Test Fixtures
const MOCK_COMMODITY_DATA = [
  { commodity: 'Soybean', variety: 'Yellow', mandi: 'Indore APMC', state: 'MP', currentPrice: 4320, avgPrice: 4250, minPrice: 4120, maxPrice: 4450, unit: '₹ / Qtl', movementPercent: 1.65, trend: 'up', historicalPrices: [{ price: 4160 }, { price: 4320 }] },
  { commodity: 'Wheat', variety: 'Sharbati', mandi: 'Sehore APMC', state: 'MP', currentPrice: 2480, avgPrice: 2510, minPrice: 2380, maxPrice: 2590, unit: '₹ / Qtl', movementPercent: -1.19, trend: 'down', historicalPrices: [{ price: 2540 }, { price: 2480 }] },
  { commodity: 'Onion', variety: 'Nashik Red', mandi: 'Lasalgaon APMC', state: 'MH', currentPrice: 1850, avgPrice: 1720, minPrice: 1540, maxPrice: 1980, unit: '₹ / Qtl', movementPercent: 7.56, trend: 'up', historicalPrices: [{ price: 1620 }, { price: 1850 }] },
  { commodity: 'Tomato', variety: 'Hybrid Table', mandi: 'Narayangaon APMC', state: 'MH', currentPrice: 1420, avgPrice: 1580, minPrice: 1350, maxPrice: 1850, unit: '₹ / Qtl', movementPercent: -10.13, trend: 'down', historicalPrices: [{ price: 1780 }, { price: 1420 }] },
  { commodity: 'Cotton', variety: 'Shankar-6', mandi: 'Rajkot APMC', state: 'GJ', currentPrice: 7150, avgPrice: 7080, minPrice: 6920, maxPrice: 7300, unit: '₹ / Qtl', movementPercent: 1.13, trend: 'up', historicalPrices: [{ price: 7020 }, { price: 7150 }] },
  { commodity: 'Maize', variety: 'Yellow Feed', mandi: 'Davangere APMC', state: 'KA', currentPrice: 2180, avgPrice: 2150, minPrice: 2050, maxPrice: 2240, unit: '₹ / Qtl', movementPercent: 1.4, trend: 'up', historicalPrices: [{ price: 2120 }, { price: 2180 }] },
];

const MOCK_SNAPSHOT_30D = {
  timeRange: '30d',
  lastUpdated: 'Today, 03:45 PM',
  spending: { totalSpend: 46200, previousPeriodSpend: 39800, percentageChange: 16.1, trendDirection: 'up', orderCount: 12, averageOrderValue: 3850 },
  spendingTrend: [{ date: '2026-08-08', label: '08 Aug', amount: 6200, orderCount: 2 }],
  categories: normalizeCategoryPercentages([
    { categoryId: 'cat_fert', categoryName: 'Fertilizers', amount: 17200, percentage: 37.2, orderCount: 4, topItem: 'NPK 10:26:26', previousPeriodAmount: 14500, growthRate: 18.6 },
    { categoryId: 'cat_prot', categoryName: 'Crop Protection', amount: 12800, percentage: 27.7, orderCount: 3, topItem: 'Coragen Bio-Pesticide', previousPeriodAmount: 8900, growthRate: 43.8 },
    { categoryId: 'cat_seeds', categoryName: 'Seeds', amount: 7400, percentage: 16.0, orderCount: 2, topItem: 'Soybean JS-335', previousPeriodAmount: 9200, growthRate: -19.5 },
    { categoryId: 'cat_irrig', categoryName: 'Irrigation', amount: 5200, percentage: 11.3, orderCount: 2, topItem: 'Drip Lateral Pipe', previousPeriodAmount: 4100, growthRate: 26.8 },
    { categoryId: 'cat_tools', categoryName: 'Farm Tools', amount: 2200, percentage: 4.8, orderCount: 1, topItem: 'Battery Sprayer', previousPeriodAmount: 1800, growthRate: 22.2 },
    { categoryId: 'cat_care', categoryName: 'Animal Care', amount: 1400, percentage: 3.0, orderCount: 1, topItem: 'Mineral Mixture', previousPeriodAmount: 1300, growthRate: 7.7 },
  ]),
  seasonalPatterns: [
    { id: 'sp_kharif_active', season: 'Kharif', primaryCategory: 'Fertilizers', peakMonths: 'July – August', observation: 'Fertilizer purchases peaked during mid-Kharif top dressing.', historicalSpend: 17200, recommendedAction: 'Plan pre-booking for Rabi DAP.', confidenceScore: 96 },
  ],
  savings: { totalSavings: 8640, productDiscountSavings: 6840, deliverySavings: 1200, bulkSavings: 600, averageDiscountPercent: 15.7, traditionalRetailEstimate: 53820, netFarmerAdvantage: 16.5, savingsByCategory: { Fertilizers: 3450, 'Crop Protection': 2680 } },
  deliveryPerformance: { totalOrders: 12, deliveredOrders: 11, activeShipments: 1, delayedDeliveries: 0, deliveryAttemptRate: 98.2, averageDeliveryHours: 41, onTimeRate: 100.0, historyTrend: [{ period: 'Week 1', onTimeRate: 100 }] },
  marketComparisons: MOCK_COMMODITY_DATA,
};

describe('Stage 9: Web Analytics, Farm Insights & Decision Intelligence', () => {
  // 1. Deterministic Calculation Utilities
  it('calculates deterministic direct savings and retail price comparisons accurately', () => {
    const totalSavings = calculateTotalSavings(6840, 1200, 600);
    assert.equal(totalSavings, 8640);

    // 16.5% offline middleman markup
    const offlineRetail = calculateRetailComparison(46200, 16.5);
    assert.equal(offlineRetail, 53823);

    const avgDiscount = calculateAverageDiscount(46200, 8640);
    assert.equal(avgDiscount, 15.8);

    const deliverySavings = calculateDeliverySavings(12, 600);
    // Baseline 12 * 150 = 1800 - 600 = 1200
    assert.equal(deliverySavings, 1200);
  });

  // 2. Spending Change & Trend Calculation
  it('computes period-over-period spending delta and trend direction', () => {
    const changeUp = calculateSpendingChange(46200, 39800);
    assert.equal(changeUp.trendDirection, 'up');
    assert.equal(changeUp.percentageChange, 16.1);

    const changeDown = calculateSpendingChange(12000, 15000);
    assert.equal(changeDown.trendDirection, 'down');
    assert.equal(changeDown.percentageChange, 20.0);

    const changeFlat = calculateSpendingChange(10000, 10020);
    assert.equal(changeFlat.trendDirection, 'flat');
    assert.equal(changeFlat.percentageChange, 0);
  });

  // 3. Category Normalization
  it('normalizes category allocation shares to 100% total', () => {
    const rawCategories = [
      { categoryId: 'c1', categoryName: 'Fertilizers', amount: 5000, percentage: 0, orderCount: 2, topItem: 'NPK', previousPeriodAmount: 4000, growthRate: 25 },
      { categoryId: 'c2', categoryName: 'Seeds', amount: 3000, percentage: 0, orderCount: 1, topItem: 'Soybean', previousPeriodAmount: 3000, growthRate: 0 },
      { categoryId: 'c3', categoryName: 'Protection', amount: 2000, percentage: 0, orderCount: 1, topItem: 'Pesticide', previousPeriodAmount: 1500, growthRate: 33 },
    ];

    const normalized = normalizeCategoryPercentages(rawCategories);
    assert.equal(normalized[0].percentage, 50.0);
    assert.equal(normalized[1].percentage, 30.0);
    assert.equal(normalized[2].percentage, 20.0);
    const sum = normalized.reduce((acc, c) => acc + c.percentage, 0);
    assert.equal(sum, 100.0);
  });

  // 4. Time Range Windowing Days
  it('converts canonical time ranges into verified windowing day spans', () => {
    assert.equal(getTimeRangeDays('7d'), 7);
    assert.equal(getTimeRangeDays('30d'), 30);
    assert.equal(getTimeRangeDays('3m'), 90);
    assert.equal(getTimeRangeDays('6m'), 180);
    assert.equal(getTimeRangeDays('1y'), 365);
  });

  // 5. Deterministic Rules-Based FarmInsightEngine
  it('evaluates deterministic insights without ungrounded heuristics', () => {
    const insights = FarmInsightEngine.evaluateSnapshot(MOCK_SNAPSHOT_30D);
    assert.ok(insights.length >= 3, 'Must detect multiple deterministic insights for 30d snapshot');

    // Verify Spending Anomaly Rule
    const anomaly = insights.find((i) => i.type === 'spending_anomaly');
    assert.ok(anomaly, 'Spending anomaly must be detected for categories exceeding 15% growth');
    assert.equal(anomaly.badgeLabel, 'Purchase Pattern');
    assert.ok(anomaly.summary.includes('fertilizers') || anomaly.summary.includes('crop protection'));

    // Verify Market Signal Rule
    const marketSignal = insights.find((i) => i.type === 'price_opportunity');
    assert.ok(marketSignal, 'Market price opportunity must be detected');
    assert.equal(marketSignal.badgeLabel, 'Market Signal');

    // Verify Delivery Performance Rule
    const deliveryInsight = insights.find((i) => i.type === 'delivery_performance');
    assert.ok(deliveryInsight, 'Delivery performance SLA compliance must be logged');
    assert.equal(deliveryInsight.badgeLabel, 'Delivery Performance');

    // Verify Seasonal Planning Rule
    const seasonalInsight = insights.find((i) => i.type === 'seasonal_reminder');
    assert.ok(seasonalInsight, 'Seasonal reminder must be triggered');
    assert.equal(seasonalInsight.badgeLabel, 'Farm Insight');
  });

  // 6. Transparent Insight Detail Completeness
  it('guarantees complete transparency diagnosis for every detected insight', () => {
    const insights = FarmInsightEngine.evaluateSnapshot(MOCK_SNAPSHOT_30D);
    insights.forEach((insight) => {
      assert.ok(insight.detail.whatHappened.length > 20, 'whatHappened must be explanatory');
      assert.ok(insight.detail.whyDetected.length > 20, 'whyDetected must explain the rule');
      assert.ok(insight.detail.supportingData.length >= 2, 'supportingData must contain verifiable rows');
      assert.ok(insight.detail.recommendedConsideration.length > 20, 'recommendation must be actionable');
    });
  });

  // 7. Commodity Market Comparison Data
  it('tracks verified APMC commodities with historical curves and high/low spreads', () => {
    assert.equal(MOCK_COMMODITY_DATA.length, 6, 'Must track all 6 institutional commodities');
    const expectedCrops = ['Soybean', 'Wheat', 'Onion', 'Tomato', 'Cotton', 'Maize'];
    expectedCrops.forEach((crop) => {
      const item = MOCK_COMMODITY_DATA.find((c) => c.commodity === crop);
      assert.ok(item, `Commodity ${crop} must be tracked`);
      assert.ok(item.historicalPrices.length >= 2, `${crop} must have historical price points`);
      assert.ok(item.maxPrice >= item.minPrice, `${crop} maxPrice must be >= minPrice`);
      assert.ok(item.avgPrice >= item.minPrice && item.avgPrice <= item.maxPrice, `${crop} avgPrice must fall in range`);
    });
  });
});
